src/esp/assets/ResourceManager.cpp

```diff
@@ -2184,5 +2185,208 @@ std::unique_ptr<MeshData> ResourceManager::createJoinedCollisionMesh(
+  Cr::Utility::Directory::writeString(filepath + "/" + new_filename,
+                                      "# CHD Mesh group \n");
+  // write vertex info to file
+  int numVertices = 0;
+  for (MeshTransformNode node : metaData.root.children) {
+    CollisionMeshData& meshData =
+        meshes_.at(node.meshIDLocal + metaData.meshIndex.first)
+            ->getCollisionMeshData();
+    for (auto& pos : meshData.positions) {
+      Cr::Utility::Directory::appendString(filepath + "/" + new_filename,
+                                           "v " + std::to_string(pos[0]) + ' ' +
+                                               std::to_string(pos[1]) + ' ' +
+                                               std::to_string(pos[2]) + '\n');
+      numVertices++;
+    }
+  }
+  Cr::Utility::Directory::appendString(
+      filepath + "/" + new_filename,
+      "# Number of vertices: " + std::to_string(numVertices) + "\n\n");
+
+  // Now do second pass to write indices for each group (node)
+  int globalVertexNum = 1;
+  int numParts = 1;
+  for (MeshTransformNode node : metaData.root.children) {
+    CollisionMeshData& meshData =
+        meshes_.at(node.meshIDLocal + metaData.meshIndex.first)
+            ->getCollisionMeshData();
+    Cr::Utility::Directory::appendString(
+        filepath + "/" + new_filename,
+        "g part_" + std::to_string(numParts) + " mesh\n");
+    for (int ix = 0; ix < meshData.indices.size(); ix += 3) {
+      Cr::Utility::Directory::appendString(
+          filepath + "/" + new_filename,
+          "f " + std::to_string(meshData.indices[ix] + globalVertexNum) + ' ' +
+              std::to_string(meshData.indices[ix + 1] + globalVertexNum) + ' ' +
+              std::to_string(meshData.indices[ix + 2] + globalVertexNum) +
+              '\n');
+    }
+    numParts++;
+    globalVertexNum += meshData.positions.size();
+  }
+
+  return success;
+}
+
+u_int ResourceManager::getNumberOfResource(const std::string& resourceName) {
+  return resourceDict_.count(resourceName);
+}
+
+void ResourceManager::createConvexHullDecomposition(
+    const std::string& filename,
+    const std::string& CHDFilename,
+    const VHACDParameters& params,
+    const bool saveCHDToObj) {
+  if (resourceDict_.count(filename) == 0) {
+    // load/check for render MeshMetaData and load assets
+    bool renderMeshSuccess =
+        loadObjectMeshDataFromFile(filename, filename, "render", true);
+
+  }  // if no render asset exists
+
+  // fill points and triangles with mesh data
+  std::vector<float> points = std::vector<float>();
+  std::vector<uint32_t> triangles = std::vector<uint32_t>();
+  getPrimitiveMeshData(filename, points, triangles);
+
+  // use VHACD
+  bool res =
+      interfaceVHACD->Compute(&points[0], (unsigned int)(points.size() / 3),
+                              (const uint32_t*)&triangles[0],
+                              (unsigned int)(triangles.size() / 3), params);
```

Comment:
Hm, since you're passing raw pointers here, is the `getPrimitiveMeshData()` function and all the data copying really necessary? The `vbo` is also a list of floats, only grouped to three-component vectors.

Same below, unless VHACD is *really badly designed* (which I don't think it is), magnum's MeshData should be able to reference the data from `ch` directly, without having to copy again. What's the type of `ch.m_points` and `ch.m_triangles`?

Reply:
Good catch, passing vbo and ibo straight in worked! I think my assumption was that a vec3f couldn't be iterated on in the same way as 3 floats, but looks like it can :)